File: demo/android/app/src/main/kotlin/com/mkparser/demo/llm/LlmProvider.kt
package com.mkparser.demo.llm

import kotlinx.coroutines.flow.Flow

/**
 * Abstraction for any text-generation backend.
 * Implement this interface to plug in a remote API (OpenAI, Ollama, …)
 * or a local on-device model.
 *
 * Usage:
 *   val provider: LlmProvider = OpenAiProvider(config)
 *   provider.stream("Write a short Markdown essay about cats").collect { token ->
 *       parser.feed(token)
 *   }
 */
interface LlmProvider {
    /** Stream tokens for the given prompt. Each emission is a raw text chunk. */
    fun stream(prompt: String): Flow<String>

    /** Human-readable label shown in the UI. */
    val label: String
}

/** Configuration for an OpenAI-compatible REST endpoint. */
data class LlmConfig(
    val baseUrl: String = "https://api.openai.com/v1",
    val apiKey:  String = "",
    val model:   String = "gpt-4o-mini",
    /** Max tokens to request. */
    val maxTokens: Int = 512,
    /** System prompt prepended to every request. */
    val systemPrompt: String = "You are a helpful assistant. Reply using Markdown formatting.",
)

File: demo/ios/App/llm/LlmProvider.swift
// LlmProvider.swift — Protocol for LLM streaming backends

import Foundation

/// Configuration for a real LLM endpoint.
struct LlmConfig {
    var apiKey:  String = ""
    var baseURL: String = "https://api.openai.com/v1"
    var model:   String = "gpt-4o-mini"
}

/// Any type that can stream Markdown text tokens from a prompt.
protocol LlmProvider {
    /// Human-readable display label (shown in the UI).
    var label: String { get }

    /// Begin streaming tokens for the given `prompt`.
    /// Calls `onToken` on the main actor for each received token.
    /// Calls `onDone` when the stream is complete or `onError` on failure.
    func stream(
        prompt:  String,
        onToken: @escaping (String) -> Void,
        onDone:  @escaping () -> Void,
        onError: @escaping (Error) -> Void
    ) -> LlmStream
}

/// An opaque handle that lets the caller cancel the stream.
final class LlmStream {
    private let cancelBlock: () -> Void
    init(cancel: @escaping () -> Void) { self.cancelBlock = cancel }
    func cancel() { cancelBlock() }
}
